File: output/python/Medium/914-longest-univalue-path.py
#https://leetcode.com/problems/longest-univalue-path/
#Given the root of a binary tree, return the length of the longest path, where each node in the path has the same value. This path may or may not pass through the root.
#The length of the path between two nodes is represented by the number of edges between them.
# 
#Example 1:
#Input: root = [5,4,5,1,1,null,5]
#Output: 2
#Explanation: The shown image shows that the longest path of the same value (i.e. 5).
#Example 2:
#Input: root = [1,4,5,4,4,null,5]
#Output: 2
#Explanation: The shown image shows that the longest path of the same value (i.e. 4).
# 
#Constraints:
#	The number of nodes in the tree is in the range [0, 104].
#	-1000 <= Node.val <= 1000
#	The depth of the tree will not exceed 1000.
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution(object):
    def longestUnivaluePath(self, root):
        """
        :type root: TreeNode
        :rtype: int
        """
        

File: ScanProject/JavaAnalysis/SonarJava/src/main/java/Medium/914-longest-univalue-path.java
package main.java.Medium; /**https://leetcode.com/problems/longest-univalue-path/ */
//Given the root of a binary tree, return the length of the longest path, where each node in the path has the same value. This path may or may not pass through the root.
//The length of the path between two nodes is represented by the number of edges between them.

class TreeNode113 {
     int val;
      TreeNode113 left;
      TreeNode113 right;
      TreeNode113() {}
      TreeNode113(int val) { this.val = val; }
      TreeNode113(int val, TreeNode113 left, TreeNode113 right) {
          this.val = val;
          this.left = left;
          this.right = right;
      }
  }

class LongestUnivaluePath {
    public int longestUnivaluePath(TreeNode113 root) {
        int[] res = new int[1];
        helper(root, res);
        return res[0];
        
    }
    public int helper(TreeNode113 root, int[] res) {
        if (root == null) {
            return 0;
        }
        int left = helper(root.left, res);
        int right = helper(root.right, res);
        int leftPath = 0;
        int rightPath = 0;
        if (root.left != null && root.left.val == root.val) {
            leftPath = left + 1;
        }
        if (root.right != null && root.right.val == root.val) {
            rightPath = right + 1;
        }
        res[0] = Math.max(res[0], leftPath + rightPath);
        return Math.max(leftPath, rightPath);
    }
}

    